hw-02-finish

Elena040683 opened the pull request.

Elena040683 committed:
hw-02-finish

KirillKulakow commented on model/index.js:
```diff
@@ -1,20 +1,61 @@
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
 
-const getContactById = async (contactId) => {}
+const listContacts = async () => {
```
почему функция асинхронная?

Elena040683 replied:
Синхронные методы плохи тем, что блокируют выполнения нашего процесса, поэтому используются в крайне редких случаях. Наш код должен быть всегда асинхронным.

KirillKulakow replied:
окей, принято
но синхронные методы не используются крайне редко, в основном асинхронные методы используются для того чтобы не блокировать стек задач (если этот метод выполняется достаточно долго или данный метод ожидает какой то ответ от другого сервиса)
В данном случае у тебя нет никаких асинхронный операций внутри данного метода - ты только отдаешь список контактов, так какой смысл - делать данный метод асинхронным?

KirillKulakow commented on model/index.js:
```diff
@@ -1,20 +1,61 @@
 
-const removeContact = async (contactId) => {}
+const getContactById = async (contactId) => {
+  const [contact] = contacts.filter((contact) => contact.id === contactId);
```
что в данном случае лучше использовать find или filter? жду аргументированный ответ

Elena040683 replied:
Если метод filter используется для поиска всех элементов, удовлетворяющих условию, то метод find позволяет найти и вернуть первый подходящий элемент, после чего перебор массива прекращается (т.е. ищет до первого совпадения, что в нашем случае подходит больше)

KirillKulakow commented on model/index.js:
```diff
@@ -1,20 +1,61 @@
+const updateContact = async (contactId, body) => {
+  const index = contacts.findIndex((contact) => contact.id === contactId);
+  if (index !== -1) {
+    const updatedContact = { id: contactId, ...contacts[index], ...body };
```
какой смысл перезаписывать ид?

Elena040683 replied:
Объект прилетает с клиента без id.

KirillKulakow replied:
ок - разберем то что ты записываешь
например у тебя следующие изначальные данные такие:
```
const contacts = [{ id: 3, name: "Mark", email: "mark@test.com", phone: "333-33-33"}, { id: 1, name: "Roman", email: "roman@test.com", phone: "444-44-44"}];
body = { name: "Rick", email: "rick@gmail.com", phone: "332-12-32" };
```
contactId пусть будет 3, в итоге мы получим:
`const updatedContact = { id: 3, ...{ id: 3, name: "Mark", email: "mark@test.com", phone: "333-33-33"}, ...{ name: "Rick", email: "rick@gmail.com", phone: "332-12-32" } };`
вопрос остается - зачем перезаписывать ид?

Elena040683 replied:
Оказывается работает и без перезаписи id

KirillKulakow reviewed (changes requested):
need fix

Elena040683 committed:
fix

Если метод filter используется для поиска всех элементов, удовлетворяющих условию, то метод find позволяет найти и вернуть первый подходящий элемент, после чего перебор массива прекращается (т.е. ищет до первого совпадения, что в нашем случае подходит больше)

KirillKulakow commented on model/index.js:
```diff
@@ -1,20 +1,61 @@
 
-const removeContact = async (contactId) => {}
+const getContactById = async (contactId) => {
+  const [contact] = contacts.find((contact) => contact.id === contactId);
```
здесь будет ошибка - подумай что возвращает метод find... если мысль не прийдет - напишешь в слак

Elena040683 replied:
Метод find возвращает первый элемент, удовлетворяющий условию, когда колбек возвращает true. Если не один элемент не подойдёт, т.е. для всех элементов колбек вернул false, метод возвращает undefined.  Find используется для одной задачи поиск элементов по уникальному значению свойства.

Elena040683 replied:
Убрала деструктуризацию.

KirillKulakow reviewed (changes requested):
Need fix

Elena040683 committed:
remove id: contactId

Elena040683 committed:
fix-contact

Убрала деструктуризацию
